C error with generics and functions

I'm continuing my ongoing testing with V generic function types (#9851, #10085).

Here is the next problem:

**V version:**  V 0.2.2 492d264
**OS:**  linux x86_64

**What did you do?**

(Note that this program is just the beginning of a real test I planned, but I hit this bug.)

```
fn neg(a int) int {
        return -a
}

struct FnHolder<T> {
        //func T // DOES NOT WORK
        func fn(int) int
}

fn (self FnHolder) call(a int) int {
        return self.func(a)
}

fn holder_call<T>(func T, a int) int {
        f := FnHolder{func}
        //return f.call(a)
        return 0
}

fn test_fn_holder() {
        mut ret := 0

        ret = holder_call(neg, 1)
        assert ret == -1
}

```


**What did you expect to see?**

No C Error.

 
**What did you see instead?**

```
==================
/tmp/v/t03_test.7304980542187936670.tmp.c:14: error: ';' expected (got "int")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -618,7 +618,7 @@ fn (mut c Checker) unwrap_generics_struct_init(struct_type ast.Type) ast.Type {
 			for i in 0 .. ts.info.generic_types.len {
 				gts := c.table.get_type_symbol(c.unwrap_generic(ts.info.generic_types[i]))
 				nrt += gts.name
-				c_nrt += gts.name
+				c_nrt += gts.cname
 				if i != ts.info.generic_types.len - 1 {
 					nrt += ','
 					c_nrt += '_'
@@ -1685,14 +1685,14 @@ fn (mut c Checker) check_return_generics_struct(return_type ast.Type, mut call_e
 	if rts.info is ast.Struct {
 		if rts.info.is_generic {
 			mut nrt := '$rts.name<'
-			mut c_nrt := '${rts.name}_T_'
+			mut c_nrt := '${rts.cname}_T_'
 			for i in 0 .. rts.info.generic_types.len {
 				if ct := c.table.resolve_generic_to_concrete(rts.info.generic_types[i],
 					generic_names, concrete_types, false)
 				{
 					gts := c.table.get_type_symbol(ct)
 					nrt += gts.name
-					c_nrt += gts.name
+					c_nrt += gts.cname
 					if i != rts.info.generic_types.len - 1 {
 						nrt += ','
 						c_nrt += '_'

--- a/vlib/v/tests/generics_struct_anon_fn_type_test.v
+++ b/vlib/v/tests/generics_struct_anon_fn_type_test.v
@@ -1,0 +1,23 @@
+fn neg(a int) int {
+	return -a
+}
+
+struct FnHolder<T> {
+	func T
+}
+
+fn (self FnHolder<T>) call<T>(a int) int {
+	return self.func(a)
+}
+
+fn holder_call<T>(func T, a int) int {
+	_ = FnHolder<T>{func}
+	return 0
+}
+
+fn test_generic_struct_with_anon_fn_parameter() {
+	mut ret := 0
+
+	ret = holder_call(neg, 1)
+	assert ret == 0
+}

